### PR TITLE
Add PKI secret backend

### PR DESCRIPTION
**Pull request by bhuisgen**
Add these resources to manage a PKI:
- `vault_pki_secret_backend`: install/configure the PKI backend
- `vault_pki_secret_backend_role`: a PKI role
- `vault_pki_secret_backend_root_cert` : generate a root CA *(do not use a root CA in Vault)*
- `vault_pki_secret_backend_root_sign_intermediate` : sign the intermediate CA by the root CA *(do not use a root CA in Vault)*
- `vault_pki_secret_backend_intermediate_cert_request` : generate intermediate CA cert
- `vault_pki_secret_backend_intermediate_set_signed` : set the intermediate with the provided signed certificate
- `vault_pki_secret_backend_cert` : generate a certificate 
- `vault_pki_secret_backend_sign`:  sign a certificate with the provided CSR
- `vault_pki_secret_backend_config_ca`: submit CA information

To test this PR:
 
```
// providers configuration

provider vault {
}

// root CA secret backend

resource "vault_pki_secret_backend" "root" {
  path                      = "test-root"
  description               = "root for test"
  default_lease_ttl_seconds = "${10 * 365 * 86400}"
  max_lease_ttl_seconds     = "${10 * 365 * 86400}"
}

// intermediate CA secret backend

resource "vault_pki_secret_backend" "intermediate" {
  path                      = "test-intermediate"
  description               = "intermediate for test"
  default_lease_ttl_seconds = "${1 * 365 * 86400}"
  max_lease_ttl_seconds     = "${1 * 365 * 86400}"
}

// generate root CA

resource "vault_pki_secret_backend_root_cert" "root" {
  backend               = "${vault_pki_secret_backend.root.path}"
  type                  = "internal"
  common_name           = "test Root CA"
  ttl                   = "${10 * 365 * 86400}"
  format                = "pem"
  private_key_format    = "der"
  key_type              = "rsa"
  key_bits              = 4096
  ou                    = "test"
  organization          = "test"
  country               = "test"
  locality              = "test"
  province              = "test"
}

// generate intermediate CSR

resource "vault_pki_secret_backend_intermediate_cert_request" "intermediate" {
  backend     = "${vault_pki_secret_backend.intermediate.path}"
  type        = "internal"
  common_name = "my.domain"
}

// sign intermediate by root

resource "vault_pki_secret_backend_root_sign_intermediate" "intermediate" {
  backend               = "${vault_pki_secret_backend.root.path}"
  csr                   = "${vault_pki_secret_backend_intermediate_cert_request.intermediate.csr}"
  common_name           = "test Intermediate CA"
  use_csr_values        = false
  permitted_dns_domains = ["my.domain"]
  ou                    = "test"
  organization          = "test"
  country               = "test"
  locality              = "test"
  province              = "test"
}

// set intermediate as signed

resource "vault_pki_secret_backend_intermediate_set_signed" "intermediate" {
  backend = "${vault_pki_secret_backend.intermediate.path}"

  certificate = "${vault_pki_secret_backend_root_sign_intermediate.intermediate.certificate}"
}

// generate a role

resource "vault_pki_secret_backend_role" "admin" {
  backend = "${vault_pki_secret_backend.intermediate.path}"

  name             = "admin"
  allowed_domains  = ["my.domain"]
  allow_subdomains = true
  max_ttl          = "${7860 * 3600}"
}

// issue a certificate

resource "vault_pki_secret_backend_cert" "test" {
  backend = "${vault_pki_secret_backend.intermediate.path}"

  name        = "${vault_pki_secret_backend_role.admin.name}"
  common_name = "test.my.domain"
}
```

### CODE REVIEW

**Comment by martinssipenko** [comment]
@bhuisgen could you update your branch with latest master in hopes that this will be bicked up soon. Having this mergen would eliminate so much manual work. Thanks for your effort!

**Comment by bhuisgen** [comment]
@tyrannosaurus-becks sync to master + fmt golang 1.11 OK

File: vault/resource_pki_secret_backend_cert.go [hunk 1]
```diff
@@ -0,0 +1,209 @@
...
+	resp, err := client.Logical().Write(path, data)
+	if err != nil {
+		return fmt.Errorf("error creating certificate %s by %s for PKI secret backend %q: %s", commonName, name,
+			backend, err)
+	}
+	log.Printf("[DEBUG] Created certificate %s by %s on PKI secret backend %q", commonName, name, backend)
+
+	d.Set("certificate", resp.Data["certificate"])
+	d.Set("issuing_ca", resp.Data["issuing_ca"])
+	d.Set("ca_chain", resp.Data["ca_chain"])
```
Comment (Erouan50) [inline vault/resource_pki_secret_backend_cert.go]: Hi @bhuisgen !

The value ca_chain returned by the API is a slice of string: https://www.vaultproject.io/api/secret/pki/index.html#sample-response-8. So with the actual code, the ca_chain will always be null. I don't know if Terraform already provides some util functions to do the conversion but here is mine if you want to use it to convert the slice of interface into a slice of string:

```hcl

d.Set("ca_chain", strings.Join(convertIntoSliceOfString(resp.Data["ca_chain"])[:], "\n"))

func convertIntoSliceOfString(slice interface{}) []string {
	intSlice := slice.([]interface{})
	strSlice := make([]string, len(intSlice))
	for i, v := range intSlice {
		strSlice[i] = v.(string)
	}
	return strSlice
}
```

**Comment by bhuisgen** [comment]
@Erouan50 fixed, thanks.

**Comment by jrasell** [comment]
@bhuisgen nice work on this; been testing it locally and there seems to be an issue with `alt_names` on the `vault_pki_secret_backend_cert` resource.

when using a list like ["ip1","ip2"] the following error is thrown:
```
error converting input [ip1 ip2] for field "alt_names": '' expected type 'string', got unconvertible type '[]interface {}'
```

when using a converted list like `${join(",", var.test)}"` where `var.test` is ["ip1", "ip2"] the following error is thrown:
```
alt_names: should be a list
```

A potential fix I tested locally works by converting the `alt_names` to a TypeString and adding it to the data map like: `"alt_names":            d.Get("alt_names").(string),`

**Comment by bhuisgen** [comment]
@jrasell please check the fix.

**Comment by tyrannosaurus-becks** [comment]
@bhuisgen thanks for updating the formatting and merging in master! The two test files are both passing, which is great. It's still needing more test coverage, though! Here's what I'm seeing:

[x] vault/resource_pki_secret_backend.go
[ ] vault/resource_pki_secret_backend_cert.go
[ ] vault/resource_pki_secret_backend_intermediate_cert_request.go
[ ] vault/resource_pki_secret_backend_intermediate_set_signed.go
[x] vault/resource_pki_secret_backend_role.go
[ ] vault/resource_pki_secret_backend_root_cert.go
[ ] vault/resource_pki_secret_backend_root_sign_intermediate.go
[ ] vault/resource_pki_secret_backend_sign.go

**Comment by bhuisgen** [comment]
Ok @tyrannosaurus-becks it's time to finish that :) I will try next days to push something.

**Comment by martinssipenko** [comment]
@bhuisgen any updates on this? I've been waiting forever to get this released.

**Comment by rmb938** [comment]
@bhuisgen Could we get this endpoint added at a resource as well? https://www.vaultproject.io/api/secret/pki/index.html#submit-ca-information

**Comment by bhuisgen** [comment]
@rmb938 it's time to check :)

**Comment by tyrannosaurus-becks** [comment]
@bhuisgen this is having the same issue as #157 was where the tests aren't runnable.

**Comment by tyrannosaurus-becks** [comment]
@bhuisgen I just re-pulled the repo and attempted to run tests again and it still didn't build. When you're ready for me to take another look at this and the other PR, just @mention me on those PR's to let me know you're ready for me to circle back. Thank you!

**Comment by bhuisgen** [comment]
@tyrannosaurus-becks latest fixes and docs added. For the failed tests you found by merging on master branch, check my other PR for an easy fix but it's a problem with a aws_secret_backend resource which should be refactored.

**Comment by zunkree** [comment]
Any ideas when this pull request will be merged?

**Comment by krish7919** [comment]
@tyrannosaurus-becks When is the next release scheduled? :)

**Pull request merged**